contentcuration/contentcuration/management/commands/garbage_collect.py

```diff
@@ -33,5 +34,7 @@ def handle(self, *args, **options):
         clean_up_deleted_chefs()
         logging.info("Cleaning up feature flags")
         clean_up_feature_flags()
+        logging.info("Cleaning up stale file objects")
+        clean_up_stale_files()
```

Comment:
This function is only cleaning up 10000 files at a time, so each run of the garbage collection job (which runs daily on production) will only clear up 10000 files - instead we should be running until there are no more files that meet the criteria. One possibility would be to return the number of deleted files, and keep on trying to delete until the number of deleted files is zero.

If we don't do this, it will take us approximately 2 years to clear all the existing 'orphaned' file objects at just 10,000 a day!

Reply:
@rtibbles It's actually 100k, not 10k. We decided to limit this because we don't know how long deleting the 7 million would take. I think what the actual limit could be adjusted. We looked at what the cron schedule of the job was in production, and a very conservative value was chosen.

Reply:
Ah, I assumed that this was intended as a batching parameter to limit the length of any particular delete query. Even with 100000 it will still be 70 days before the backlog is cleared.

Is there a specific concern with the scheduled job taking a long time? It seems like it will be a one time long run and then after that will just be keeping on top of it.

Reply:
We saw that the cron jobs are scheduled to run every 24 hours. While looking at Django’s delete() method, we noticed that it also triggers the deletion signals, which could take some time to process, along with the memory usage of deleting millions of file objects in one go. The potential for the job to run past 24 hours was our main concern for the limit.

Reply:
There seem to be two concerns here - one about limiting the memory usage and time of the deletion operation, and one about the run time of the cronjob.

I think we should be limiting the memory usage, and could even reduce the batch size further to remove any concerns.

For the cronjob, I think it is more important that it runs to completion than it finish within 24 hours - as this is parallel to the other deletions that are happening in the cronjob. Without ensuring that all the file objects matching this criteria are deleted, we continue to run the risk of a file table filled with an ever expanding number of 'junk' files.

In terms of concerns about signals, there is one signal that is currently attached to the post_delete signal of the File object, which recalculates file storage: https://github.com/learningequality/studio/blob/4444843712b7dad06528ac38b62a2f15b5dc7486/contentcuration/contentcuration/models.py#L2232 - you are already handling this using the decorator that delays storage recalculation.

Reply:
> Without ensuring that all the file objects matching this criteria are deleted, we continue to run the risk of a file table filled with an ever expanding number of 'junk' files.

@rtibbles So you're saying we could get more than 100k files in 24 hours?

> I think it is more important that it runs to completion than it finish within 24 hours

If it doesn't finish in 24 hours, couldn't the subsequent run deadlock with the previous run?

Reply:
Increasing the limit seems like a good compromise

Reply:
> I don't know that it would deadlock, maybe the subsequent or existing run would end up running superfluous deletes?

> At the moment, I think this is plausible when running the KA EN Chef (although I am trying to fix the ricecooker behaviour to prevent this).

Since there is quite a bit of unknown here, seems like it would be safer to limit it then. So how about a limit of 500k? Or 1 million?

Reply:
I am fine with a higher limit - do we still want to batch though to limit memory usage? Presumably for the `post_delete` signal the full model representation is being loaded, and I'm not sure if Django is batching that or not.

Reply:
Seems like it doesn't do any batching, but it does check if there are any signal handlers to see whether it needs to load the model representations - this might explain why the other delete methods for garbage collection are explicitly disabling the signal handlers with a context manager.

https://docs.djangoproject.com/en/3.2/ref/models/querysets/#delete

Reply:
So if we are going to increase the limit, we probably need to disable the signal handlers, rather than just deferring the user storage calculation.

However, it seems like we don't actually need to run the user storage calculation at all, as by definition, these files are not part of any active trees (and the user storage calculation only calculates for active trees).

Reply:
Got it, I will update the function to include a higher limit and remove the storage calculation. Do you have any good examples and/or links for batching?